### PR TITLE
Enable use_io_binding = True on CPU

### PR DESCRIPTION
**Pull request by yihonglyu**
# What does this PR do?

KV cache is large for large prompt (~1GB in total for prompt 1024) but onnxruntime python binding copies all outputs from ORT memory to Python memory. It is resolved with IO binding. With this patch, we can use code like
```
model = ORTModelForCausalLM.from_pretrained(
    model_path,
    use_io_binding = True,
)
```
to enable IO binding with CPU

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

**Comment by fxmarty** [comment]
I indeed see a slight speedup, but I tested with a model not really large (gpt2-large, 1000 token context length, batch size = 1, about ~7% speedup). I am kind of surprised we would see any speedup because of this: https://github.com/microsoft/onnxruntime/issues/10000#issuecomment-991220117 

It could be that there is some kind of overhead in the non-IO Binding path in Optimum that is not present in the IOBinding one.

What kind of speedup do you see?

**Comment by fxmarty** [comment]
Hi @tianleiwu, do you have any thoughts on this?

**Comment by tianleiwu** [comment]
@fxmarty, 

IO Binding for CPU can help performance when input/output tensor is huge (like past_key_value for long sequence length). Without IO Binding, the interop of python and ONNX Runtime need copy of tensor (between numpy tensor and OrtValue), and memory copy is slow when tensor size is large.

File: optimum/onnxruntime/base.py [hunk 1]
```diff
@@ -311,7 +311,7 @@ def forward(
             input_ids, past_key_values, use_torch=use_torch
         )
 
-        if self.device.type == "cuda" and self.parent_model.use_io_binding:
+        if (self.device.type == "cuda" or self.device.type == "cpu") and self.parent_model.use_io_binding:
```
Comment (tianleiwu) [inline optimum/onnxruntime/base.py:314]: For CPU, it does not need synchronize_inputs() synchronize_outputs in line 343 below.

**Comment by fxmarty** [comment]
Thank you for the details @tianleiwu!

File: optimum/onnxruntime/base.py (hunk 1 above)
Comment (yihonglyu) [reply optimum/onnxruntime/base.py:314]: Exclude synchronize_inputs() for CPU

File: optimum/onnxruntime/utils.py [hunk 2]
```diff
@@ -252,7 +252,7 @@ def check_io_binding(providers: List[str], use_io_binding: Optional[bool] = None
     """
     if use_io_binding is None and providers[0] == "CUDAExecutionProvider":
         use_io_binding = True
-    elif providers[0] != "CUDAExecutionProvider":
+    elif providers[0] != "CPUExecutionProvider" and providers[0] != "CUDAExecutionProvider":
         if use_io_binding is True:
             logger.warning(
                 "No need to enable IO Binding if the provider used is not CUDAExecutionProvider. IO Binding will be turned off."
```
Comment (tianleiwu) [inline optimum/onnxruntime/utils.py:258]: update message to include CPUExecutionProvider

Comment (yihonglyu) [reply optimum/onnxruntime/utils.py]: Updated the message.

**Review by tianleiwu — APPROVED** [review]

**Comment by yihonglyu** [comment]
@fxmarty Could you give me some review comments?

**Comment by fxmarty** [comment]
For some reason I have shape errors with batch size > 1 with CPU EP + IO Binding with this PR, while fine on CUDA EP. Will investigate and get back to you shortly.

**Review by fxmarty — APPROVED** [review]
LGTM, actually made it the default for ORTModelForCausalLM. Could be made default for encoder-decoder models as well.

**Pull request merged**